keps/sig-apps/2307-job-tracking-wihout-lingering-pods/README.md

```diff
@@ -0,0 +1,599 @@
+
+1. The new implementation produces new API calls: one per Pod to remove
+   finalizers and one for each Job sync, to track the intermediate state.
+   However, the ability to handle higher number of Pods and Jobs at a given
+   point justifies the added cost of the new API calls.
+2. Changes in the status not produced by the Job controller in
+   kube-controller-manager could affect the Job tracking. Cluster administrators
+   should make sure to protect the Job status endpoint via RBAC.
+
+## Design Details
+
+### API changes
+
+The Job status gets a new struct to hold the uncounted Pods before they are
+added to the counters.
+
+```golang
+type JobStatus struct {
+    Succeeded int32
+    Failed    int32
+    ...
+
+    // UncountedPodUIDs holds UIDs of Pods that have finished but haven't
+    // been accounted in the counters.
+    UncountedPodUIDs UncountedPodUIDs
+}
+
+// UncountedPodUIDs holds UIDs of Pods that have finished but haven't been
+// accounted in Job status counters.
+type UncountedPodUIDs struct {
+    // Enabled indicates whether Job tracking uses this struct.
+    Enabled   bool
+    // Succeeded holds UIDs of succeeded Pods.
+    Succeeded []string
+    // Succeeded holds UIDs of failed Pods.
+    Failed    []string
+}
+```
+
+### Algorithm
+
+The following algorithm updates the status counters without relying on finished
+Pods to be present indefinitely. The algorithm assumes that the Job controller
+could be stopped at any point and executed again from the first step without
+losing information. Generally, all the steps happen in a single Job sync
+cycle.
+
+1. The Job controller creates Pods for a Job with the finalizer
+   `batch.kubernetes.io/job-completion`
+2. The Job controller adds Pod UIDs to the `.status.uncountedPodUIDs.succeeded`
+   and `.status.uncountedPodUIDs.failed` lists if the Pod:
+    - has the `batch.kubernetes.io/job-completion` finalizer, and
+    - the Pod is on Succeeded or Failed phase, respectively.
+    The controller sends a status update.
+3. The Job controller removes the `batch.kubernetes.io/job-completion` finalizer
+   from all Pods on Succeeded or Failed phase that were added to the lists in
+   `.status.uncountedPodUIDs` in the previous step. The controller keeps note
+   of the Pod updates that succedeed.
+4. The Job controller counts the Pods in the `.status.uncountedPodUIDs` lists
+   that:
+   - had no finalizer at the beginning of the sync, or
+   - had the patch in step 3 succeed, or
+   - were removed from the system.
+   The counts increment the `.status.failed` and `.status.succeeded` and clears
+    counted Pods from `.status.uncountedPodsUIDs` lists. The controller sends a
+    status update.
+
+Steps 2 to 4 might deal with a potentially big number of Pods. Thus, status
+updates can potentially stress the kube-apiserver. For this reason, the Job
+controller repeats steps 2 to 4, dealing with up to 100 Pods each time, until
+the controller processes all the Pods.
```

Comment:
I think you added this in response to Daniel's comment.
@lavalamp , were you worried about too frequent of updates (not batching) or about the size of the diff from a single update (which the above seems concerned with).

Reply:
I don't think I'm concerned about either. I do think it's probably good to limit the size of the chunks we process just to guarantee progress (partial update is better than never completing a huge update). But I personally would probably have a time limit-- process all the pods you can in 1m. I would say 15s but we have to consider the possibility of apiserver taking the maximum amount of time to respond to a request (currently 60s). The tests probably need to test under the "apiserver is really slow" condition.

Reply:
> process all the pods you can in 1m.

Uhmm. That would have to look like:
- repeatedly update K Pods in parallel while 1m hasn't passed.
- setup a channel and K workers, continue processing for 1m.

So then we have to decide on K. Which exact method we use can probably be left for code reviews.

Reply:
cc @wojtek-t since you were asking about batching updates in the other KEP

Reply:
Hmm - I don't understand why updating the status (a) with N pods and (b) with M pods is different from kube-apiserver perspective.

I think what we should do is: (a) take a time into account, (b) have a cap on number of pods addeded.

We need a limit to ensure that the object won't grow to super large case, e.g. if job-controller was down for some time and in the meantime gazillions pod finished.

Reply:
So I take this as: we need to cap on time and number. Updated the KEP